[SYCLomatic] Make device_vector AllocatorAwareContainer

danhoeflinger opened the pull request:
Changes to make `dpct::device_vector` allocator aware, while still relying upon a USM shared memory allocator.

This should satisfy the requirements of [AllocatorAwareContainer](https://en.cppreference.com/w/cpp/named_req/AllocatorAwareContainer).  

With the underlying data being host accessible, we can fully satisfy AllocatorAwareContainer.

Tested by https://github.com/oneapi-src/SYCLomatic-test/pull/314.

*Rebased after .inc removal*

tomflinda reviewed (approved):
LGTM

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:
```diff
@@ -121,143 +122,142 @@ private:
+    }
+    _size = std::move(other._size);
+    _capacity = std::move(other._capacity);
+    _storage = std::move(other._storage);
```
If `std::allocator_traits<Allocator>::propagate_on_container_move_assignment::value` is false, we need use current `alloc` to allocate storage and deallocate `other`'s storage.

danhoeflinger replied:
Yes, you are correct.  I'll make this and the other related changes requested below.  I have missed some aspects of destruction / deallocation, and will update accordingly.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:
```diff
@@ -275,28 +275,32 @@ public:
     ::std::swap(_storage, v._storage);
-    ::std::swap(_alloc, v._alloc);
+    if constexpr (::std::allocator_traits<
+                      Allocator>::propagate_on_container_swap::value) {
```
Similar as above.

danhoeflinger replied:
I've addressed this section as well.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:
```diff
@@ -121,143 +122,142 @@ private:
-                other.begin(), other.end(), begin());
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_copy_assignment::value) {
+      _alloc = other._alloc;
```
Should we deallocate current storage here?

danhoeflinger replied:
Yes, done.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:
```diff
@@ -121,143 +122,142 @@ private:
-    this->swap(dummy);
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_move_assignment::value) {
+      _alloc = std::move(other._alloc);
```
Should we deallocate current storage here?

danhoeflinger replied:
Yes, done.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:
```diff
@@ -121,143 +122,142 @@ private:
+
+  void _construct(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      for (int i = 0; i < n; i++) {
```
Can we speed up the construction with dpl functions?

danhoeflinger replied:
I'm not sure we can assume the thread safety of construct or destroy calls.  

If we could assume this thread safety, from a technical perspective we cant send these construct calls to a SYCL kernel because of the use of variadic functions.  However, if we could be sure that the construction are thread safe we could either use a `par_unseq` policy or create our own version of `std::allocator_traits` with a fixed number of possible arguments to the construct call to avoid the variadic function call allow us to use them in SYCL kernels.  I have a prototype of something like this from a different implementation of this code.

danhoeflinger reviewed (commented):
Thanks for the comments.  I've addressed them here in this PR, but I've marked this as a draft for now while I upgrade the testing to cover more of the code.  
Constructors and assignment operators handling of allocators needs to be tested better (especially the different scenarios of allocator settings for propagation).  Once I am finished upgrading the testing, I will convert this back to be a ready to merge PR.
In the meantime, I welcome more comments.

mmichel11 commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,206 @@ class device_vector {
-  device_vector()
-      : _alloc(get_default_queue()), _size(0), _capacity(_min_capacity()) {
+
+  device_vector(const Allocator &alloc)
```
Can we give this a default argument, so we don't need the additional constructor in line 125?

danhoeflinger replied:
Yes, addressed in a few places.

mmichel11 commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,206 @@ class device_vector {
+
+  explicit device_vector(size_type n) : device_vector(n, get_default_queue()) {}
+
+  explicit device_vector(size_type n, const T &value, const Allocator &alloc)
```
Can we remove the need for the separate constructor in line 146 by having the constructor's signature be `explicit device_vector(size_type n, const T &value, const Allocator &alloc = get_default_queue())`

danhoeflinger replied:
Yes, I fixed these and a few others, using `const Allocator &alloc = Allocator(get_default_queue())`.

mmichel11 commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,206 @@ class device_vector {
+      : device_vector(v.begin(), v.end()) {}
+
+  template <typename OtherAllocator>
+  device_vector(::std::vector<T, OtherAllocator> &v, const Allocator &alloc)
```
Can `v` be taken as const reference here?

danhoeflinger replied:
Yes,  fixed.

mmichel11 commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,206 @@ class device_vector {
+      clear();
+      ::std::allocator_traits<Allocator>::deallocate(_alloc, _storage,
+                                                     _capacity);
+      _alloc = std::move(other._alloc);
```
`std::` -> `::std::` if we want to be consistent with the convention we use.

danhoeflinger replied:
Thanks, fixed this and others in the PR.

mmichel11 reviewed (approved):
LGTM

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,191 @@ class device_vector {
   void _set_capacity_and_alloc() {
     _capacity = ::std::max(_size * 2, _min_capacity());
-    _storage = _alloc.allocate(_capacity);
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
```
Using an alias for `::std::allocator_traits<Allocator>::` may simplify the code.

danhoeflinger replied:
Thanks, done.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,191 @@ class device_vector {
-                first, last, begin());
-    }
+  device_vector(device_vector &&other, const Allocator &alloc)
+      : _alloc(alloc), _size(other.size()), _capacity(other.capacity()),
```
If `other.alloc` is unequal to `alloc`, the `other.storage` can not reuse.

danhoeflinger replied:
You are correct, fixed.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -258,61 +337,75 @@ class device_vector {
   void shrink_to_fit(void) {
     if (_size != capacity()) {
       size_type tmp_capacity = ::std::max(_size, _min_capacity());
```
If current capacity equal to `_min_capaticy()`, the following code can be jumped.

danhoeflinger replied:
Thanks, fixed.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,191 @@ class device_vector {
     resize(v.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
+    ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
We may use the queue of allocator for the dpl algorithm.

danhoeflinger replied:
I don't believe there is a way to extract the `queue` information from the allocator.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -258,61 +337,75 @@ class device_vector {
     if (_size > 0) {
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                begin(), begin() + n, x);
+      ::std::fill(
```
Should we do `clear()` to destroy the objects at first?

danhoeflinger replied:
We are following the convention of `std::vector`, which I believe `assign` does not destroy the previous elements, it merely copies in the value or incoming elements into the vector.

ziranzha replied:
So what will the dpl algorithm do when it tries to replace existed element? Do the inplace construct or assignment?

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -258,61 +337,75 @@ class device_vector {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                first, last, begin());
+      ::std::copy(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
Same as above.

danhoeflinger replied:
We are following the convention of `std::vector`, which I believe `assign` does not destroy the previous elements, it merely copies in the value or incoming elements into the vector.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,143 +67,191 @@ class device_vector {
 public:
   template <typename OtherA> operator ::std::vector<T, OtherA>() const {
     auto __tmp = ::std::vector<T, OtherA>(this->size());
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
Here the objects construction is speeded up by dpl algorithm, but you comments said that is not make sense. So which is expected?

danhoeflinger replied:
This instance is copying elements, not constructing them.  It does not have the issues described below.

The issue with construction being sped up by a SYCL kernel or oneDPL algorithm is two-fold: 
1) Conceptually, allocators (which may be custom user-supplied overrides of `usm_allocator`) can't be assumed to have thread safe `construct` calls.  For instance, if you had a constructor which counted the number of element constructions and did not provide synchronization around the count variable, this could have issues.
2) Technically, `std::allocator_traits<T>::construct()` is a variadic function, and those cannot be used SYCL device functions.  Therefore, it cannot be called within a SYCL kernel.  As I mention before, if parallel accellerated construction is a feature which is highly requested, we can create our own version of `allocator_traits` which does not rely on variadic functions for use in this context.  However, with (1) in mind, it does not seem worthwhile to add this complexity.  

If we add this, we would need to state some rules about what is allowed in custom `construct` calls.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -258,61 +337,75 @@ class device_vector {
-    auto tmp = _alloc.allocate(m);
+    auto tmp = ::std::allocator_traits<Allocator>::allocate(_alloc, m);
     // copy remainder to temporary buffer.
     ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
Why copy to a temporary buffer? Just move ahead is much simpler.

danhoeflinger replied:
This is to take advantage of the parallel algorithms, as input and output sequences may overlap if we just move ahead.

I suppose we can check if they overlap and choose which method we use accordingly.  I've done this.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -325,57 +418,62 @@ class device_vector {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                position, end(), tmp);
+      ::std::copy(
+          oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
Similar as above.

danhoeflinger replied:
See above comment.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -325,57 +418,62 @@ class device_vector {
       auto m = ::std::distance(position, end());
       // will throw if position is not inside active vector
-      auto tmp = _alloc.allocate(m);
+      auto tmp = ::std::allocator_traits<Allocator>::allocate(_alloc, m);
```
Why it will throw?

danhoeflinger replied:
It will throw because it would be attempting to allocate a negative number of bits for the temporary.

ziranzha replied:
Is it possible that the the distance is a positive number and not to large, so the allocation can be done?

danhoeflinger replied:
Yes, in normal operation where this position *is* inside the vector, it the distance will return a reasonable positive number and the allocation will succeed.

danhoeflinger reviewed (commented):
Thanks for the feedback, I've replied and made some changes in the code

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -217,31 +259,60 @@ class device_vector {
-      _alloc.deallocate(_storage, _capacity);
+      if (capacity() > 0) {
+        ::std::copy(
+            oneapi::dpl::execution::make_device_policy(get_default_queue()),
```
I think here we can not use dpl to speed up, as construction is doing here.

danhoeflinger replied:
We are performing a copy to move existing (previously constructed) data to a new memory location, not calling the constructor here.  I don't believe any constructor is called here, and thus, using oneDPL to accelerate the copy is appropriate.

ziranzha replied:
The new memory is not initialized, for some complex type, it may cause unexpected errors when doing copy assignment (which I think will be done inside `std::copy`).

danhoeflinger replied:
Sorry for the delayed response here.  After doing some more experimentation with std::vector and a custom allocator, you are correct, and this use of oneDPL for speeding up this operation is dubious.  

However, I'm worried that fixing this will result in punishing the common case to account for the rare case of a custom allocator's custom constructor.  

I'm working on a idea which will use parallel `std::uninitialized_copy` when no custom constructor is specified by the allocator, but will use serial calls to the allocators custom constructor when one is specified.  (the default `usm_allocator` does not define one).  If I can get it to work, I'll take a sweep of the PR here to make sure all similar instances are handled appropriately.  

Thanks for bringing this up.

AndyCHHuang reviewed (approved):
LGTM
According to the test result of small test example of std copy, I think std copy does not trigger copy constructor.

AndyCHHuang commented:
@danhoeflinger 
Please rebase the PR,

danhoeflinger commented:
@ziranzha  I've made some changes which should hopefully address your concerns about copying to uninitialized memory, and using oneDPL acceleration to construct memory.  

I've added the struct `device_allocator_traits<Allocator>` which acts somewhat similarly to `std::iterator_traits`, but at a slightly higher level.  If a matching custom constructor is specified by the allocator, then we sequentially call that custom constructor on all the data we are constructing, filling, copying to, etc.  This is because we have no guarantees on thread safety / device safety of such custom constructors.  

However, we do not want to slow down the normal case (without custom constructor from allocator, likely TriviallyCopyable types), for the rare case with a custom constructor.  Therefore, If no matching custom constructor is provided by the supplied allocator, then we send it to the relevant `uninitialized_fill`, `uninitialized_copy`, `uninitialized_default_construct_n` routines with a parallel execution policy.  These routines should properly handle uninitialized data.  

To pair with this I've also fixed in a few places where the constructor was incorrectly being skipped (in favor of a std::copy).  These places now correctly use the `_construct` routines, which utilize the new `device_allocator_traits<Allocator>` struct.

Please take a look and see if this resolves your concerns.  

*edit* I see a couple failures in the tests,  I'll need to look at these tomorrow.

tomflinda commented:
@danhoeflinger  the latest CI test failure below is introduced by this PR.
```
thrust_scatter_if RUNFAIL 
```
Pls fix this regression test before merging this PR.

danhoeflinger commented:
@tomflinda @mmichel11 @ziranzha I believe all issues have been addressed.  Tests are passing (most recent change is formatting only but retriggered tests).  

Take a look and merge if you are happy.

mmichel11 commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -41,6 +42,139 @@ struct is_iterator<T *> : std::true_type {};
+  template <typename T, typename Size>
+  static
+      typename ::std::enable_if_t<__has_destroy<_Allocator, T *>::value, void>
+      destroy_n(_Allocator alloc, T *p, Size n) {
```
Do we want some assert to check that `p != nullptr` throughout these utilities or is that unnecessary here?

danhoeflinger replied:
Good idea, added these asserts.

mmichel11 reviewed (commented):
One minor question from me about the added device allocator traits utility.

mmichel11 reviewed (approved):
The recent changes LGTM. Reapproving.

danhoeflinger commented:
@zhimingwang36 @tomflinda I think this is ready to merge.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -41,6 +42,150 @@ struct is_iterator<T *> : std::true_type {};
+  struct __has_construct_impl : ::std::false_type {};
+
+  template <class _Alloc, class... _Args>
+  struct __has_construct_impl<decltype((void)std::declval<_Alloc>().construct(
```
I think we can use `__has_construct` directly here, like `__has_destroy`.

danhoeflinger replied:
I don't believe its possible to use directly because of the variable number of arguments `construct()` can accept.  The reason this works for `__has_destroy` gets to be simpler, is that we know that it only accepts a single pointer as an argument.  

They are effectively using the same trick, but because of the template variable pack, we have to utilize the additional template argument at the beginning and use the `_impl` layer rather than at the end of the template parameters with a defaulted argument.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -41,6 +42,150 @@ struct is_iterator<T *> : std::true_type {};
+  template <typename T, typename Size>
+  static
+      typename ::std::enable_if_t<__has_construct<_Allocator, T *>::value, void>
+      uninitialized_value_construct_n(_Allocator alloc, T *p, Size n) {
```
How about implement `uninitialized_value_construct_n`, `uninitialized_fill_n`, `uninitialized_device_copy_n` inside `__has_construct` to make code cleaner and simpler?

danhoeflinger replied:
I'm not sure I understand this request.  How would one implement these inside of `__has_construct`?

One thing I could think that _might_ make things simpler is to convert the `::std::enable_if_t<__has_construct<_Allocator, T *>::value, void>` with different functions to `if constexpr(__has_construct<_Allocator, T *>::value)` within each of these `uninitialized_...` functions.  This would consolidate each function to a single instance, and could consolidate currently repeated assertions.  Is this what you might be asking for?

danhoeflinger replied:
I did what I mentioned here above and I think I prefer it, so I kept it.  Let me know if you had something else in mind.

ziranzha replied:
LGTM

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,145 +213,189 @@ class device_vector {
+      _alloc = other._alloc;
+    }
+    reserve(other.capacity());
     resize(other.size());
```
How about use `uninitialized_copy` instead of `resize` and `copy`?

danhoeflinger replied:
Yes, I think this can be improved, but we have a few different scenarios to cover.  
We need special cases for 
1) when the incoming size is <= existing size
`std::copy`, then destroy excess

2) when it isn't but the incoming size is <= _capacity
`std::copy_n` the already constructed elements, `unintialized_device_copy_n` the rest, then set new size

3) when incoming elements exceed current capacity
destroy all existing elements, `uninitialized_device_copy_n` all elements, set size

danhoeflinger replied:
fixed, thanks.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,145 +213,189 @@ class device_vector {
+      _size = ::std::move(other._size);
+    } else {
+      reserve(other.capacity());
+      resize(other.size());
```
Same as above.

danhoeflinger replied:
Basically the same can be done as described above.

danhoeflinger replied:
fixed, thanks.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -219,31 +410,58 @@ class device_vector {
+      // then copy the elements beyond the end of the smaller list, and resize
+      if (size() > v.size()) {
+        size_type tmp_v_size = v.size();
+        v.resize(size());
```
Same as above.

danhoeflinger replied:
Thanks for catching this, it needs a bit of work.  We probably need special casing for different capacity vs size situations.  We also may want a special swapping parallel for in order to avoid bringing the data to the host to swap it, and to extra temporary allocation requirements.
Uninitialized fill makes sense for the excess.

danhoeflinger replied:
fixed, thanks.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -327,57 +560,59 @@ class device_vector {
       resize(size() + n);
-      ::std::fill(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                end() - n, end(), x);
+      ::std::fill(
```
How about use `uninitialized_fill` instead of `std::fill` and `resize`?

danhoeflinger replied:
Yes, this should be fixed.  Thanks.

danhoeflinger replied:
fixed, thanks.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,145 +213,189 @@ class device_vector {
-    }
+    // unsafe to parallelize on device as we dont know if InputIterator is valid
+    // oneDPL input type
+    _construct_host(first, last);
```
If user defined allocator is used and the allocated memory can only be accessed by device, it will not workable. Should we consider such cases?

danhoeflinger replied:
We don't support those cases.  The underlying premise which enables `device_vector` to work is that we are relying on shared USM memory.  If we move away from that, it no longer works generally.

The USM_none case relies on `sycl::buffers`, (and we don't support the same functionality in that case).

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -66,145 +213,189 @@ class device_vector {
+
+  template <typename Iter>
+  void _construct(Iter first, Iter last, size_type start_idx = 0) {
+    size_type num_eles = ::std::distance(first, last);
```
I think we can accept size as the input argument to avoid call of `std::distance` since the size is known in use positions.

danhoeflinger replied:
edit: my original response was wrong.  

Sure, we can make this change.

danhoeflinger replied:
fixed, thanks.

ziranzha commented on clang/runtime/dpct-rt/include/dpl_extras/vector.h:
```diff
@@ -326,58 +548,60 @@ class device_vector {
       resize(size() + n);
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                first, last, end());
+      // unsafe to call on device as we dont know the InputIterator type
```
I think we can also use `uninitialized_copy_n` here.

danhoeflinger replied:
done.  I looked at changing some of the other `resize` to `reserve` and `uninitialized_x`, but the others are not clear wins like this one seems to be.

ziranzha reviewed (approved):
LGTM
